src/array.jl

```diff
@@ -1050,3 +1050,27 @@ StructTypes.construct(::Type{<:CategoricalArray{Union{Nothing, T}}},
                       A::Vector) where {T} =
     categoricalnothing(T, A)
 categoricalnothing(T, A::AbstractVector) = CategoricalArray{Union{Nothing, T}}(A)
+
+# DataAPI refarray/refvalue/refpool support
+struct CategoricalRefPool{T, P} <: AbstractVector{T}
+    pool::P
+end
+
+Base.IndexStyle(::Type{<: CategoricalRefPool}) = Base.IndexLinear()
+@inline function Base.getindex(x::CategoricalRefPool, i::Int)
+    @boundscheck checkbounds(x, i)
+    i > 0 ? @inbounds(x.pool[i]) : missing
+end
+Base.size(x::CategoricalRefPool{T}) where {T} = (length(x.pool) + (T >: Missing),)
+Base.axes(x::CategoricalRefPool{T}) where {T} =
+    ((T >: Missing ? 0 : 1):length(x.pool),)
+Base.LinearIndices(x::CategoricalRefPool) = axes(x, 1)
```

Comment:
:clap:

Reply:
Can you explain what this is needed for? I'm not familiar with LinearIndices. Does it have to do with defining `axes` as well? Or why is that needed?

Reply:
This is needed because by default `Base.LinearIndices` would return `1:length(x)` even if the valid indices are from `0` to `length(x)-1` incorrectly.

It is not likely that in normal use cases anyone would use `Base.LinearIndices` but who knows what users will try to do.